File: packages/cexplorer-api/src/endpoints/tool.ts
import type { TxSentResponse } from "@/types/toolTypes";

import { handleFetch } from "@/lib/handleFetch";

/**
 * Sends delegation or donation transaction information to track campaign participation.
 *
 * This endpoint is typically used to notify the backend about successful delegation
 * or donation transactions, allowing for campaign tracking and analytics.
 *
 * @param {string} hash - The transaction hash.
 * @param {string} poolId - The pool ID or campaign identifier.
 * @param {"delegation" | "donate"} [type="delegation"] - The type of transaction (delegation or donation).
 *
 * @returns {Promise<TxSentResponse>} A promise resolving to the transaction sent confirmation.
 *
 * @throws Will throw an error if the API request fails.
 */
export const sendTxSent = async (
  hash: string,
  poolId: string,
  type: "delegation" | "donate" = "delegation",
) => {
  const url = "/tool/tx_sent";
  const options = {
    params: {
      id: hash,
      type,
      campaign: poolId,
    },
  };

  return handleFetch<TxSentResponse>(url, undefined, options);
};

File: packages/cexplorer-api-rs/src/endpoints/tool.rs
use crate::client::fetch_with_params;
use crate::error::CexplorerError;
use crate::types::tool_types::*;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct TxSentParams {
    pub id: String,
    pub r#type: String,
    pub campaign: String,
}

/// Send delegation or donation transaction information
pub async fn send_tx_sent(
    hash: &str,
    pool_id: &str,
    tx_type: &str,
) -> Result<TxSentResponse, CexplorerError> {
    let endpoint = "/tool/tx_sent";
    let params = TxSentParams {
        id: hash.to_string(),
        r#type: tx_type.to_string(),
        campaign: pool_id.to_string(),
    };
    fetch_with_params::<TxSentResponse, TxSentParams>(endpoint, Some(&params)).await
}
